datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+
+        ds = require_dataset(
+            dataset, check_installed=True, purpose='pushing')
+
+        if since:
+            # will blow with ValueError if unusable
+            ds.repo.get_hexsha(since)
+
+        if not since and since is not None:
+            # TODO figure out state of remote branch and set `since`
+            pass
+
+        # obtain a generator for information on the datasets to process
+        # idea is to turn the `paths` argument into per-dataset
+        # content listings that can be acted upon
+        if since:
+            ds_spec = _datasets_since_(
+                # important to pass unchanged dataset arg
+                dataset, since, paths, recursive, recursion_limit)
+        else:
+            ds_spec = _datasets_no_since_(
+                # important to pass unchanged dataset arg
+                dataset, paths, recursive, recursion_limit)
+
+        res_kwargs = dict(
+            action='publish',
+            refds=ds.path,
+            logger=lgr,
+        )
+        # instead of a loop, this could all be done in parallel
+        matched_anything = False
+        for dspath, dsrecords in ds_spec:
+            matched_anything = True
+            lgr.debug('Attempt push of Dataset at %s', dspath)
+            yield from _push(
+                dspath, dsrecords, to, force, jobs, res_kwargs.copy())
+        if not matched_anything:
+            yield dict(
+                res_kwargs,
+                status='notneeded',
+                message='Given constraints did not match any changes to publish',
+                type='dataset',
+                path=ds.path,
+            )
+
+
+def _datasets_since_(dataset, since, paths, recursive, recursion_limit):
+    """Generator"""
+    # rely on diff() reporting sequentially across datasets
+    cur_ds = None
+    ds_res = None
+    for res in diff_dataset(
+            dataset=dataset,
+            fr=since,
+            # we never touch unsaved content
+            to='HEAD',
+            constant_refs=False,
+            path=paths,
+            # we need to know what is around locally to be able
+            # to report something that should have been pushed
+            # but could not, because we don't have a copy
+            annex='availability',
+            recursive=recursive,
+            recursion_limit=recursion_limit,
+            # make it as fast as possible
+            eval_file_type=False,
+            # we rely on all records of a dataset coming out
+            # in succession, with no interuption by records
+            # concerning subdataset content
+            reporting_order='breadth-first'):
+        if res.get('action', None) != 'diff':
+            # we don't care right now
+            continue
+        if res.get('status', None) != 'ok':
+            # we cannot handle this situation, report it in panic
+            raise RuntimeError(
+                'Cannot handle non-OK diff result: {}'.format(res))
```

Comment:
It seems like doing a hard stop here is in conflict with the "instead of a loop, this could all be done in parallel" comment at the call site.  Is it not?

Reply:
That is true. However, this condition and also the one tested afterwards should not happen (they are more like asserts). If they happen something is wrong, not just for the dataset in question, but with the logic/assumptions in general. I think it is fine to stay like this.

Reply:
Fair enough.